### ISSUE TITLE
feature: cell border

### ISSUE THREAD

**Issue by scanny** [scanny/python-pptx#71]
set borders of table cell; left, right, top, and bottom

Attributes: color (fill), thickness (weight), style (single, thick-thin, thin-thin, etc.), inset pen (inside boundary or centered on it), line join (round, bevel, miter), and line cap (round, square, flat).

**Comment by glayman** [comment]
Hi, are you still planning to implement this feature? If not, is there a way that I can incorporate this functionality until a feature is built? For now, I need the ability to draw horizontal borders across multiple cells in a table, and set the color, weight and style.

**Comment by smcpherson** [comment]
👍  for this feature request.

**Comment by rpeys** [comment]
I would benefit from this feature as well! If it won't be implemented for a while, have any hints for how to access this xml feature?

**Comment by dereitz** [comment]
Are there any workarounds to this?  I've run into a scenario where I'd like to change the border in a table.  Can the default be changed such that no borders are introduced in a table (as a workaround)?

**Comment by dereitz** [comment]
I found this thread.  One thing to note is that order appears to be important.  In my case, the lnB element had to come before the solidFill element!

https://groups.google.com/forum/#!topic/python-pptx/UTkdemIZICw

**Comment by ahuang11** [comment]
A reproducible example; as dereitz mentioned, make sure you set the `cell.fill.solid()` **after** borders!
```
import pptx
from pptx import Presentation
from pptx.util import Inches
from pptx.oxml.xmlchemy import OxmlElement

def SubElement(parent, tagname, **kwargs):
    element = OxmlElement(tagname)
    element.attrib.update(kwargs)
    parent.append(element)
    return element


def _set_cell_border(cell, border_color="000000", border_width='12700'):
    """ Hack function to enable the setting of border width and border color
        - bottom border only at present
        (c) Steve Canny
    """
    tc = cell._tc
    tcPr = tc.get_or_add_tcPr()

    lnR = SubElement(
        tcPr, 'a:lnR', w=border_width, cap='flat', cmpd='sng', algn='ctr')
    solidFill = SubElement(lnR, 'a:solidFill')
    srgbClr = SubElement(solidFill, 'a:srgbClr', val=border_color)
    return cell



prs = Presentation()
slide = prs.slides.add_slide(prs.slide_layouts[5])
placeholder = slide.placeholders[0]
x, y, cx, cy = Inches(2), Inches(2), Inches(4), Inches(1.5)
shape = slide.shapes.add_table(3, 3, x, y, cx, cy)
table = shape.table

for r in range(3):
    for c in range(3):
        cell = table.cell(r, c)
        cell = _set_cell_border(cell)
        cell.fill.solid()
        cell.fill.fore_color.rgb = pptx.dml.color.RGBColor(255, 0, 0)
        if c == 1:
            break
prs.save('test.pptx')
```

**Comment by MartinPacker** [comment]
If this were ever to make it into the python-pptx code I'd also like to enable / disable table borders. I see the `<tblborder>` element in the spec. (See [here](http://officeopenxml.com/WPtableBorders.php).)

For [md2pptx](https://github.com/martinpacker/md2pptx) I will probably roll my own.

It just seems to me table and cell borders really would be a consistent thing to work on together.

**Comment by MartinPacker** [comment]
Actually that link in my previous only works for Excel. Hex dumping a table with a border doesn't reveal how it does it.

**Comment by scanny** [comment]
@MartinPacker judging from how the PowerPoint UI allows borders to be applied I'd say that's on a cell-by-cell basis, like to put a border around a table you have to visit each perimeter cell and set its outward-facing borders to the desired value.

If I select the table and click on "Table Design > Borders > Outside Borders" then PowerPoint applies those individually to each perimeter cell. Interestingly, if I then add a new row at the bottom by tabbing through the last cell, it carries those borders down in the expected way. So it must have some "border-moving" logic in there on growing or shrinking the table.

It's possible such a thing can be set using a customized table style but it's not immediately clear to me how after a look at the schema. I expect there is a series of experiments between here and knowing with confidence what can be managed there.

**Comment by MartinPacker** [comment]
Right. That's what I gathered. And that implies one of two approaches:

1. Apply the border after all the cells have been created.
2. Adjust the border settings on adding cells.

If I do this I think I can manage to do Option 1 in md2pptx - as I know the numbers of rows and columns when creating the table. But I can see in general this wouldn't be fun for python-pptx to do.

**Comment by ahuang11** [comment]
I found that code here so I'm not sure:
https://groups.google.com/g/python-pptx/c/UTkdemIZICw

**Comment by scanny** [comment]
No, there's no copyright. @ahuang11 Just a URL to the mailing-list source is fine if you want to add a citation, but I'm not particular about that. Best not to add a copyright though. Anything I post here or on SO or a mailing-list is public domain as far as I'm concerned.

**Comment by Lusipher** [comment]
Hello everyone!!
The `_set_cell_border` hack in this thread works mostly — but it only cleanly handles uniform borders, and it falls apart the moment adjacent cells need different colors (PowerPoint makes them fight over the shared edge).
 
I ran into this enough that I packaged a workaround: **pptx-extended** (`pip install pptx-extended`). It wraps `set_cell_border()` for all four sides + dash styles, and adds a **GappedTable _(experimental)_** that uses thinnest possible cells (I call them  _spacer cells_ ;D) so adjacent cells can have genuinely independent borders. Here's a screenshot of how the table would look:

<img width="960" height="517" alt="Image" src="https://github.com/user-attachments/assets/46659cf5-7ba7-4116-b5d3-6aa21dbeb256" />

I wrote this library — sharing because this exact thread is what sent me down the rabbit hole.
Repo: https://github.com/bms-pub-rbit/pptx-extended

**Comment by MartinPacker** [comment]
Ooh! I like the idea of "gap cells". It gets me tiles. I might add this into md2pptx - as a slide-level option.
